File: MoogFilter.m
classdef MoogFilter < AudioEffect
   
    properties
        
        numberOfStages
        temperatureConstant
        
        cutoffFrequencyInput
        resonanceInput
        
    end
    
    methods
        
        function this = MoogFilter(name, numberOfStages)
    
            this = this@AudioEffect(name);
            this.numberOfStages = numberOfStages;
            this.temperatureConstant = 1;
                        
            this.cutoffFrequencyInput = this.createInputPort( 0.5 );
            this.resonanceInput = this.createInputPort( 0.1 );
            
        end;
        
        function updateFx(this, N)
            
            [x, numberOfChannels] = this.input.read(N);
            
            FcN = this.cutoffFrequencyInput.read(N);
            r = this.resonanceInput.read(N);
            
            Vt = this.temperatureConstant;
            noStages = this.numberOfStages;
            
            ys = zeros(1,noStages);  % Ladder filter stages

            y = zeros(size(x));      % Filter output;

            for( c = 1:numberOfChannels )
                
                for( n = 1:N )

                    g = 1 - exp(-2*pi*FcN(n)/2);

                    stageInput = x(n,c) - 4*r(n)*ys(noStages);   % Input signal and feedback from last stage

                    for( k = 1:noStages )

                        ys(k) = ys(k) + 2*Vt*g*( tanh(stageInput/(2*Vt)) - tanh(ys(k)/(2*Vt)) );

                        stageInput = ys(k);

                    end;

                    y(n,c) = ys(noStages);
                    
                end;
                
            end;
            
            this.output.write( y );
            
        end;
        
    end
end

File: Cpp/Modules/Source/MoogFilter.cpp
#include "MoogFilter.h"
#include <corecrt_math_defines.h>

MoogFilter::MoogFilter()
	: _stageOutputs()
{
	_pFrequencyInput = CreateInputPort();
	_pResonanceInput = CreateInputPort();

	_pFrequencyInput->Set(0.5);
	_pResonanceInput->Set(0.1);
}

IInputPort* MoogFilter::GetFrequencyInput() const
{
	return _pFrequencyInput;
}

IInputPort* MoogFilter::GetResonanceInput() const
{
	return _pResonanceInput;
}

double MoogFilter::Update(double x)
{
	static const auto VT = 1.5; // Temperature constant

	const auto g = 1 - exp(-2 * M_PI * _pFrequencyInput->Read() / 2);

	auto stageInput = x - 4 * _pResonanceInput->Read() * _stageOutputs[NUMBER_OF_STAGES - 1]; // Input signal and feedback from last stage

	for (auto& stageOutput : _stageOutputs)
	{
		stageOutput = stageOutput + 2 * VT * g * (tanh(stageInput / (2 * VT)) - tanh(stageOutput / (2 * VT)));

		stageInput = stageOutput;
	}

	return _stageOutputs[NUMBER_OF_STAGES - 1];
}
